BPA/Ch.10/js_rest_occ.stan

```diff
@@ -0,0 +1,232 @@
+   *
+   * @return Integer value of last capture occasion
+   */
+  int last_capture(int[] y_i) {
+    for (k_rev in 0:(size(y_i) - 1)) {
+      int k;
+      k <- size(y_i) - k_rev;
+      if (y_i[k])
+        return k;
+    }
+    return 0;
+  }
+
+  /**
+   * Return a matrix of uncaptured probability
+   *
+   * @param nind        Number of individuals
+   * @param n_occasions Number of capture occasions
+   * @param p           Detection probability for each individual
+   *                    and capture occasion
+   * @param phi         Survival probability for each individual
+   *                    and capture occasion
+   *
+   * @return Uncaptured probability matrix
+   */
+  matrix prob_uncaptured(int nind, int n_occasions,
+                         matrix p, matrix phi) {
+    matrix[nind, n_occasions] chi;
+
+    for (i in 1:nind) {
+      chi[i, n_occasions] <- 1.0;
+      for (t in 1:(n_occasions - 1)) {
+        int t_curr;
+        int t_next;
+
+        t_curr <- n_occasions - t;
+        t_next <- t_curr + 1;
+        chi[i, t_curr] <- (1 - phi[i, t_curr]) +
+                          phi[i, t_curr] *
+                          (1 - p[i, t_next]) *
+                          chi[i, t_next];
+      }
+    }
+    return chi;
+  }
+}
+
+data {
+  int<lower=0> M;                              // Augmented sample size
+  int<lower=0> n_occasions;                    // Number of capture occasions
+  int<lower=0,upper=1> y[M, n_occasions];      // Augmented capture-history
+}
+
+transformed data {
+  int<lower=0,upper=n_occasions> first[M];
+  int<lower=0,upper=n_occasions> last[M];
+
+  for (i in 1:M)
+    first[i] <- first_capture(y[i]);
+  for (i in 1:M)
+    last[i] <- last_capture(y[i]);
+}
+
+parameters {
+  real<lower=0,upper=1> mean_phi;              // Mean survival probability
+  real<lower=0,upper=1> mean_p;                // Mean capture probability
+  vector<lower=0,upper=1>[n_occasions] gamma;  // Removal entry probability
+}
+
+transformed parameters {
+  matrix<lower=0,upper=1>[M, n_occasions - 1] phi;  // Survival
+  matrix<lower=0,upper=1>[M, n_occasions] p;        // Capture
+  matrix<lower=0,upper=1>[M, n_occasions] chi;
+
+  // Constraints
+  for (i in 1:M) {
+    for (t in 1:(n_occasions - 1))
+      phi[i, t] <- mean_phi;
+    for (t in 1:n_occasions)
+      p[i, t] <- mean_p;
+  } //i
+
+  // Uncapture probability
+  chi <- prob_uncaptured(M, n_occasions, p, phi);
+}
+
+model {
+  vector[n_occasions] qgamma;
+
+  qgamma <- 1.0 - gamma;
+
+  // Priors
+  mean_phi ~ uniform(0, 1);
+  mean_p ~ uniform(0, 1);
+  gamma ~ uniform(0, 1);
+
+  // Likelihood
+  for (i in 1:M) {
+    vector[n_occasions] qp;
+
+    qp <- 1.0 - p[i]';
+
+    if (first[i]) { // Captured
+      // Until first capture
+      if (first[i] == 1) {
+        1 ~ bernoulli(gamma[1] * p[i, 1]);
+      } else {
+        vector[first[i]] lp;
+
+        // Entered at 1st occasion
+        lp[1] <- bernoulli_log(1, gamma[1]) +
```

Comment:
I think it's clearer with addition symbols starting the line (you can't do that in R, but you can in Stan).